### ISSUE TITLE
Keras LSTM Language Model using Embeddings

### ISSUE THREAD

**Issue by florent1989** [fchollet/keras#3587]
Hi

I am doing a language model using keras.

Basically, my vocabulary size N is ~30.000, I already trained a word2vec on it, so I use the embeddings, followed by LSTM, and then I predict the next word with a fully connected layer followed by softmax. My model is written as below :
`EMBEDDING_DIM = 256
embedding_layer = Embedding(N,EMBEDDING_DIM,weights=[embeddings],
trainable=False)

model = Sequential()
model.add(embedding_layer)
model.add(LSTM(EMBEDDING_DIM))
model.add(Dense(N))  
model.add(Activation('softmax')) 

model.compile(loss="categorical_crossentropy", optimizer="rmsprop")`

I have two questions :
1. What do you think, instead of connecting the last hidden layer of the lstm to a big fully connected layer of size N (30.000), connecting to a layer of size EMBEDDING_DIM, and predicting the embedding of the next word instead of the word itself, in which case we replace the loss by something like mse, reducing training time, and mainly "helping" our model because the vocabulary is big and embeddings can be useful also for the end of the network ?
2. In this case, can you confirm that we only use the last hidden layer of the LSTM (which is followed by the fully connected layer and softmax) and there isn't something like a max/mean-pooling of successive hidden layers of the lstm (like here for sentiment analysis http://deeplearning.net/tutorial/lstm.html) ?

Thanks !

**Comment by lomizandtyd** [comment]
Hi, florent, I'm working on language model too...

-----
Prob 1.   

- Something like  "hierarchical softmax (in word2vec)" , or " class models" are classical way to deal with large vocabulary size. 
- The optimized target of mse is different from ce. I never use "word2vec" as true representation of a word.  Try to image a special case, the embedding size of a word is 128, if you predict a word, embedding features No. 1~128 is exactly right, but feature No. 128 is totally wrong, which lead the result to another word. You would get low mse but higher cross-entropy in such case.     

Prob 2.

- Max / Mean Pooling could be done by use a Lambda layer.

-----

Other 
About how I implement my language model..

I use [w1, w2, w3, ..., wN]  as input of a sentence,
the output  is [w2, w3, w4, ..., wN, sentenceEnd]
